network/src/peers/peer/outbound_handler.rs

```diff
@@ -0,0 +1,162 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::time::Instant;
+
+use tokio::sync::{mpsc, oneshot};
+
+use crate::{stats, NetworkError, Payload, Peer};
+
+use super::network::PeerNetwork;
+
+pub type PeerJudge = dyn Fn(&Peer) -> bool + Send + Sync + 'static;
+
+pub(super) enum PeerAction {
+    Disconnect,
+    Send(Payload),
+    Get(oneshot::Sender<Peer>),
+    QualityJudgement,
+    CancelSync,
+    GotSyncBlock(oneshot::Sender<bool>),
+    ExpectingSyncBlocks(u32),
+    SoftFail,
+}
+
+#[derive(Clone, Debug)]
+pub struct PeerHandle {
+    pub(super) sender: mpsc::Sender<PeerAction>,
+}
+
+impl PeerHandle {
+    pub async fn load(&self) -> Option<Peer> {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        let (sender, receiver) = oneshot::channel();
+        self.sender.send(PeerAction::Get(sender)).await.ok()?;
+        receiver.await.ok()
+    }
+
+    pub async fn judge(&self) {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        self.sender.send(PeerAction::QualityJudgement).await.ok();
+    }
+
+    /// returns true if disconnected, false if not connected anymore
+    pub async fn disconnect(&self) -> bool {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        self.sender.send(PeerAction::Disconnect).await.is_ok()
+    }
+
+    pub async fn send_payload(&self, payload: Payload) {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        self.sender.send(PeerAction::Send(payload)).await.ok();
+    }
+
+    pub async fn cancel_sync(&self) {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        self.sender.send(PeerAction::CancelSync).await.ok();
+    }
+
+    pub async fn got_sync_block(&self) -> bool {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        let (sender, receiver) = oneshot::channel();
+        if self.sender.send(PeerAction::GotSyncBlock(sender)).await.is_err() {
+            return false;
+        }
+        receiver.await.unwrap_or(false)
+    }
+
+    pub async fn expecting_sync_blocks(&self, amount: u32) {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        self.sender.send(PeerAction::ExpectingSyncBlocks(amount)).await.ok();
+    }
+
+    pub async fn soft_fail(&self) {
+        metrics::increment_gauge!(stats::QUEUES_OUTBOUND, 1.0);
+        self.sender.send(PeerAction::SoftFail).await.ok();
+    }
+}
+
+pub(super) enum PeerResponse {
```

Comment:
this name is a bit misleading, as it might suggest a network response; maybe a `PeerHandlerResponse` or something similar would work better

Reply:
i disagree, this abstraction layer is separate from IO entirely, it's adjacent to `PeerAction`, and is local anyways.

Reply:
it's just a matter of increased clarity - we do call other nodes "peers" after all, so if someone was looking at this part of the code without being aware of the whole picture, they might be confused